### PR TITLE
SOLR-16796: Add Maven SBOMs via cyclonedx

### PR DESCRIPTION
**Pull request by raboof**
This draft PR shows one way to generate an SBOM for Solr, mostly for discussion at https://lists.apache.org/thread/j35clzm48s7xxc9671qogzr54bsdj1lz . If we want to take this further I'll create a JIRA issue and fill out the rest of the PR template ;).

(I haven't looked in depth at the quality of the output yet, as at this point I just wanted a starting point for experimenting with VEX)

https://issues.apache.org/jira/browse/SOLR-16796

# Description

It would be nice if Solr published an 'SBOM' (Software Bill of Materials) for its artifacts. An SBOM gives an overview of the components included in the artifact, which can be useful for example for scanner software that looks for dependencies with potential security vulnerabilities.

Such consumers of the SBOM should probably combine it with the VEX published for Solr (https://solr.apache.org/security.html#vex) to avoid getting reports for known false positives.

# Solution

Introduce the org.cyclonedx.bom plugin which helps with generating SBOM's for the build products.

# Checklist

Please review the following and check all that apply:

- [ ] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [ ] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [ ] I have developed this patch against the `main` branch.
- [ ] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -28,6 +28,19 @@ plugins {
...
+  includeConfigs = ["runtimeClasspath"]
+  skipConfigs = ["compileClasspath", "testCompileClasspath"]
+  projectType = "application"
+  schemaVersion = "1.4"
+  destination = file("build/reports")
+  outputName = "CycloneDX-Sbom"
+  outputFormat = "all"
+  includeBomSerialNumber = true
+  componentVersion = "2.0.0"
 }
```
Comment (risdenk) [inline build.gradle:44]: This should end up in a separate config file like `gradle/validation/cyclonedx.gradle` - another example of similar config would be any of the files in `gradle/validation`

**Review by risdenk — CHANGES_REQUESTED** [review]

**Comment by risdenk** [comment]
@raboof I pushed a change to your PR branch that moves the logic to `gradle/validation/cyclonedx-bom.gradle` to match other build plugins. As far as I can tell this is a decent first step.

**Comment by epugh** [comment]
@gus-asf you may find this idea interesting ;-)

**Comment by HoustonPutman** [comment]
We should create a JIRA for this btw.

**Comment by raboof** [comment]
> We should create a JIRA for this btw.

Created https://issues.apache.org/jira/browse/SOLR-16796

There's probably more to improve from here, such as:
* Generating SBOMs 'per artifact' rather than one big one for all of Solr?
* Publishing nightly boms to https://security-tools-ec2-va.apache.org/
* Publishing the SBOM as a release artifact

.. but this might already be a nice start.

**Review by risdenk — APPROVED** [review]

File: gradle/validation/cyclonedx-bom.gradle [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    skipConfigs = ["compileClasspath", "testCompileClasspath"]
+    projectType = "application"
+    schemaVersion = "1.4"
+    destination = file("build/reports")
+    outputName = "cyclonedx-bom"
+    outputFormat = "all"
+    includeBomSerialNumber = true
+  }
+
+  // Unless explicitly enabled, do not attach owasp to check. It has a large download
```
Comment (dsmiley) [inline gradle/validation/cyclonedx-bom.gradle:34]: What does CycloneDX have to do with OWASP?  What checking/verification does `cyclonedxBom` do?

Comment (ErnestoLeon87) [reply gradle/validation/cyclonedx-bom.gradle:34]: It helps you to build a bom.xml/json file so you can evaluate that file later on Sonatype Lifecycle and see your app vulnerabilities.

**Comment by janhoy** [comment]
I'm positive to including this as a first step and then proceeding with publishing SBOM as a release artifact as proposed.

**Comment by anshumg** [comment]
Thanks @HoustonPutman . I think this would be a good first step considering we've been stuck with this for a while.

**Review by anshumg — APPROVED** [review]

**Pull request merged**